django/contrib/admindocs/utils.py

```diff
@@ -151,90 +154,185 @@ def replace_metacharacters(pattern):
+    group_pattern_and_name = [(pattern[start:end], group_name)
+                              for start, end, group_name in group_start_end_indices_and_name]
     for group_pattern, group_name in group_pattern_and_name:
         pattern = pattern.replace(group_pattern, group_name)
     return pattern
 
 
 def replace_unnamed_groups(pattern):
-    r"""
-    Find unnamed groups in `pattern` and replace them with '<var>'. E.g.,
-    1. ^(?P<a>\w+)/b/(\w+)$ ==> ^(?P<a>\w+)/b/<var>$
-    2. ^(?P<a>\w+)/b/((x|y)\w+)$ ==> ^(?P<a>\w+)/b/<var>$
-    3. ^(?P<a>\w+)/b/(\w+) ==> ^(?P<a>\w+)/b/<var>
-    4. ^(?P<a>\w+)/b/((x|y)\w+) ==> ^(?P<a>\w+)/b/<var>
-    """
-    unnamed_group_indices = [m.start(0) for m in unnamed_group_matcher.finditer(pattern)]
-    # Indices of the start of unnamed capture groups.
-    group_indices = []
-    # Loop over the start indices of the groups.
-    for start in unnamed_group_indices:
-        # Handle nested parentheses, e.g. '^b/((x|y)\w+)$'.
-        unmatched_open_brackets, prev_char = 1, None
-        for idx, val in enumerate(pattern[start + 1:]):
-            # Check for unescaped `(` and `)`. They mark the start and end of
-            # a nested group.
-            if val == '(' and prev_char != '\\':
-                unmatched_open_brackets += 1
-            elif val == ')' and prev_char != '\\':
-                unmatched_open_brackets -= 1
-            prev_char = val
+    """Finds unnamed groups in the given `pattern` and replaces them with
+    `<var>`.
 
-            if unmatched_open_brackets == 0:
-                group_indices.append((start, start + 2 + idx))
-                break
-    # Remove unnamed group matches inside other unnamed capture groups.
-    group_start_end_indices = []
-    prev_end = None
-    for start, end in group_indices:
-        if prev_end and start > prev_end or not prev_end:
-            group_start_end_indices.append((start, end))
+        * ^(?P<a>\w+)/b/(\w+)$ ==> ^(?P<a>\w+)/b/<var>$
+        * ^(?P<a>\w+)/b/((x|y)\w+)$ ==> ^(?P<a>\w+)/b/<var>$
+        * ^(?P<a>\w+)/b/(\w+) ==> ^(?P<a>\w+)/b/<var>
+        * ^(?P<a>\w+)/b/((x|y)\w+) ==> ^(?P<a>\w+)/b/<var>
+
+    Finds the string for unnamed capture groups inside the given `pattern` and
+    replaces all the sub-strings for unnamed capture groups inside pattern with
+    the name `<var>`.
+
+    Args:
+        pattern: `Regex` pattern for a `named` group.
+
+    Returns:
+        `pattern` string where unnamed group sub-strings are all replaced with
+        just the name `<var>`.
+    """
+    group_start_end_indices = _find_groups(pattern, unnamed_group_matcher, False)
+    if not group_start_end_indices:
+        return pattern
+    final_pattern, prev_end = '', None
+    for start, end in group_start_end_indices:
+        if prev_end:
+            final_pattern += pattern[prev_end:start]
+        final_pattern += pattern[:start] + '<var>'
```

Comment:
@felixxm I removed the calls to `append()` and `join()` method because I don't think we should cost two stack frames for both of these methods when we can just use a slicing operation 🤔.